### ISSUE TITLE
What is the best way to save blobs? Sometimes blob isn't returned correctly. 

### ISSUE THREAD

**Issue by trackedsupport** [dfahlander/Dexie.js#1305]
I record video at 15 second intervals and then put the video blob in the db. For some people/users/browsers the blob_data is very large but once it goes into the db, it comes out as 15 bytes. Anyone know why? 

```

   db.version(1).stores({ blobs: "++id,recording_id,filename,blob_data,attendee_id" });

   db.blobs.put({ filename: index+".webm", blob_data: blob, recording_id: recording_id, attendee_id: attendee_id}).then (function(record){
       console.log(`inserted ${index+".webm"} into the local db`)
     }).catch(function(error) {
       console.error ("ERROR SAVING BLOB: " + error);
     });

  var blob = await db.blobs.where({recording_id: recording_id}).first(1)
```

I then decided to change the blob_data to save as an ArrayBuffer with 

`var blob_data = await new Response(blob).arrayBuffer();` 

But that seems to cause a bunch of `InvalidStateError: Failed to execute 'transaction' on 'IDBDatabase': The database connection is closing. ` issues. And I think its causing `OpenFailedError: QuotaExceededError Encountered full disk while opening backing store for indexedDB.open.` errors. 

What is the best way to save blobs?

**Comment by dfahlander** [comment]
Blobs are the best way to store videos and images. Do not list `blob_data` among indexes. It won't hurt because IndexedDB will ignore it since it doesn't support indexing blobs. However, if you choose to store it as ArrayBuffer, you will get huge problems if you index the ArrayBuffer property as indexes are expensive for large values and are not meant for anything else than short keys to use in where- or orderBy queries.

Seems the problem could be that your user is out of storage quota. If you are developing a PWA, check if you can use [StorageManager](https://dexie.org/docs/StorageManager) to manage and opt-in for more storage quota.

Blob support has also been a bit buggy in older versions of Safari. Quota varies among browsers and devices. What browser, device and version is it on the users having issues?

**Comment by trackedsupport** [comment]
The latest person running into the issue has Firefox 88 Windows. If they use Chrome then it works. 

I have seen a person using Chrome having the same issue though. 

So are you saying I should not use the array buffer? I don't believe I'm creating any indexes. I might put it back to just storing blobs because of the amount of `The database connection is closing. InvalidStateError ` errors im getting. The bad thing is that then ill get the original issue for some people. Maybe I can check and store as array buffer if storing as blob doesnt work.

**Comment by dfahlander** [comment]
You list blob_data in your index list of the 'blobs' table which is unnecessary and problematic if you put an ArrayBuffer there.

**Comment by trackedsupport** [comment]
Ah okay, I guess I need to learn some more about db creation. 

If I don’t do an index, and use ArrayBuffer, will that be fine? Or do you still recommend just saving the blob? I want to do it as recommended. I just need to figure out that other 15 byte issue. Unfortunately I can’t reproduce but I do see it and did a video call with a customer and saw it happening.

**Comment by trackedsupport** [comment]
I just had the 15 byte issue happen again. This time for Chrome v91 on Catalina. Surely something is wrong! Is there something wrong with my code?!

create the db

```
    var db = new Dexie("Recordings");
    db.version(1).stores({ recordings: "++id,recording_id" });
    db.version(1).stores({ blobs: "++id,recording_id,filename,blob_data,attendee_id" });
    global.db = db
```

media recorder saving blobs


```
recorder = new MediaRecorder(local_media_stream, {
           mimeType: encoding_options,
           audioBitsPerSecond: 96000,
           videoBitsPerSecond: bits_per_second,
         });
         recorder.ondataavailable = function(e) {
             that.save_blob(e.data, blob_index);
             blob_index++;
           }
         }
```

writing the blobs in the save_blob function (if I check here the blob length is large)

```
db.blobs.put({ filename: index+".webm", blob_data: blob, recording_id: recording_id, attendee_id: attendee_id}).then (function(record){
       console.log(`inserted ${index+".webm"} into the local db`)
     }).catch(function(error) {
       console.error("ERROR SAVING BLOB1: " + error);
     });
```

upload blob (if I check here the blob length is 15 bytes)

```
var blob = await db.blobs.where({recording_id: recording_id}).first(1)
     var filename     = blob["filename"]
     var blob_data    = blob["blob_data"]
     var attendee_id  = blob["attendee_id"]
     var recording_id = blob["recording_id"]
     var file = new File([blob_data], filename, {lastModified: new Date().getTime(), type: blob_data.type});
     $('.directUpload').find("input:file").fileupload('add', {files: [file]});

```

**Comment by dfahlander** [comment]
just by looking briefly:
1. You list version 1 two times per table. This is ok though, but a bit different.
2. You still index "blob_data" in the "blobs" table. Remove it. You never know what kind of bug this could trigger in the browser, and as said before, if you store a large ArrayBuffer instead of a blob, the index will definitely give you performance problems.
3. Collection.first() doesn't take a number argument.
4. Try are you storing recording_id as an index in the recording table? Looks more like primary key to me.
5. Make sure you retrieve the correct blob in your query. Given that you only pick the first match, what if you have several blobs with the same recording_id?

**Comment by trackedsupport** [comment]
Thanks for the helpful info!

1. & 2. So I should do it like this to correct and update everyone's DB?
```
    var db = new Dexie("Recordings");
    db.version(2).stores({ 
      recordings: "++id,recording_id" , 
      blobs: "++id,recording_id,filename,attendee_id" 
    });
```
3. Ill remove it. 
4. You are right. 
5. None of the blobs should be that empty byte one. 

So if I dont have the index on blob_data and Im still running into the empty byte blob issue when I pull it out of Dexie, do you recommend I switch to ArrayBuffer then? I know that for one person switching to ArrayBuffer solved the problem.

**Comment by dfahlander** [comment]
Yes, by incrementing version to 2 and removing blob_data, the index will be gone, which makes it possible to switch to using ArrayBuffer if you believe that could be a solution. Also, when this 15 byte issue occurs again, log the result of calling `await db.blobs.where({recording_id: recording_id}).count()` - just to verify you haven't duplicate rows on the non-unique index recording_id. And please share with us the end solution when you find on it ;)
